packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -381,13 +401,33 @@ export class TokenDetectionController extends StaticIntervalPollingController<
     }, this.getIntervalLength());
   }
 
-  #getCorrectChainId(networkClientId?: NetworkClientId) {
-    const { chainId } =
-      this.messagingSystem.call(
+  #getCorrectChainIdAndNetworkClientId(networkClientId?: NetworkClientId): {
+    chainId: Hex;
+    networkClientId: NetworkClientId;
+  } {
+    if (networkClientId) {
+      const networkConfiguration = this.messagingSystem.call(
         'NetworkController:getNetworkConfigurationByNetworkClientId',
-        networkClientId ?? this.#networkClientId,
-      ) ?? {};
-    return chainId ?? this.#chainId;
+        networkClientId,
+      );
+      if (networkConfiguration) {
+        return {
+          chainId: networkConfiguration.chainId,
+          networkClientId,
+        };
+      }
+    }
+    const { chainId } = this.messagingSystem.call(
+      'NetworkController:getProviderConfig',
+    );
+    const newNetworkClientId = this.messagingSystem.call(
+      'NetworkController:findNetworkClientIdByChainId',
```

Comment:
I'm trying to fix a related bug here: https://github.com/MetaMask/core/pull/3927

I can't tell if this PR would also be affected by the bug, and it doesn't have to block this PR.  But can you take a peek and see if it makes sense to you?